geode-core/src/main/java/org/apache/geode/distributed/internal/MaxLongGauge.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.distributed.internal;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.geode.Statistics;
+
+/**
+ * This class holds the max value of a stat inside an AtomicLong. Every time a higher value is found
+ * the class will forward the delta between the old max and the new max to the statistics class.
+ * This pattern is a lock-less alternative to calling Statistics.getLong and then updating the max.
+ */
+class MaxLongGauge {
+  private final int statId;
+  private final Statistics stats;
+  private final AtomicLong max;
+
+  public MaxLongGauge(int statId, Statistics stats) {
+    this.statId = statId;
+    this.stats = stats;
+    max = new AtomicLong();
+  }
+
+  public void recordMax(long currentValue) {
+    boolean done = false;
+    while (!done) {
+      long maxValue = max.get();
+      if (currentValue <= maxValue) {
+        done = true;
+      } else {
+        done = max.compareAndSet(maxValue, currentValue);
+        if (done) {
+          stats.incLong(statId, currentValue - maxValue);
```

Comment:
why not `stats.setLong(statId, currentValue)`?

Reply:
That was how I coded it up originally but it has a race in which the value we set in the stat long may not be the max. If two threads are concurrently calling this code and both try to change it, t1 is changing 0 to 1 and and t2 is changing 0 to 2. If t1 wins the atomic check it proceeds and will call setLong on the stat. But it could lose the cpu, t2 fails the compareAndSet loops around sets the atomic to 2 and calls setLong(2). Now t1 wakes up and sets it to 1. So the atomic is correct but the stat is wrong. Murtuza's idea was to use incLong instead of setLong. The atomic is used to compute the delta and we just change the stat by that amount. The other nice thing is incs on a LongAdder perform much better than sets. Do you see a logic error with this code?

Reply:
Gotcha, yeah, it's fine. My other comment is an alternative way to solve this by just not counting a separate stat at all.